### DISCUSSION TITLE
RLS and CLS together example.

### DISCUSSION THREAD

**Discussion by mehmetuken** [supabase/supabase#3158] [Questions]
I want row level security and column level security together using. But i can't do it.

Example:
```tsql
create table todos
(
    id          uuid                     default uuid_generate_v4(),
    user_id     uuid references auth.users                                    not null,
    inserted_at timestamp with time zone DEFAULT timezone('utc'::text, now()) NOT NULL,
    updated_at  timestamp with time zone DEFAULT timezone('utc'::text, now()) NOT NULL,
    name        text,
    readonly_data text,
    primary key (id)
);
alter table todos enable row level security;
create policy "Can only view own todos." on todos for select using (auth.uid() = user_id);
create policy "Can only create own todo." on todos for insert with check(auth.uid() = user_id);
create policy "Can only update own todo." on todos for update using (auth.uid() = user_id);
create policy "Can only delete own todo." on todos for delete using (auth.uid() = user_id);

REVOKE UPDATE ("readonly_data") ON todos FROM supabase_admin,public, anon, authenticated;
```
I wan't readonly_data can't update rls auth user. I trying but user update all time.

**Comment by silentworks** [comment]
I would suggest using a view in this case or rethinking your schema modelling. If you use a view then you can only allow certain columns. If you rethink your schema modelling then you could move readonly_data to a different table and only allow `SELECT` queries on that table with RLS.

**Reply by mehmetuken** [reply]
my question is why not working this method. I wonder why it's not working.

**Reply by steve-chavez** [reply]
> I wonder why it's not working.

How are you making the request? And is there an error message?

**Reply by mehmetuken** [reply]
```js
await supabase
    .from("todos")
    .update({ readonly_data: "dummy" })
    .eq("id", id)
```

no error it's working and update readonly_data field. but i'm revoke supabase_admin,public, anon, authenticated; field to update. That's way very helpful everyone, don't need write every table function to update or insert.

**Reply by selbyk** [reply]
@silentworks How about columns like created_at and updated at? They exist on practically every table and shouldn't be user updated. Do we need a separate metadata table to store this properly for every other table? Is the performance there to be able to do this join for every query?

**Reply by abhay187** [reply]
Good point raised @selbyk if you are doing anything critical such as showing a post or comment based on created_at or updated_at Then a malicious user can update it and change the result.
Right now  I am following a metadata table approach on a table where I don't want the user to change certain columns.
Yes, you will be able to do this join for every query. For better performance try indexing foreign keys in case you are not satisfied.

**Comment by steve-chavez** [comment]
@mehmetuken Can you try doing:

```sql
alter default privileges in schema public revoke all on tables from supabase_admin,public, anon, authenticated;
alter default privileges in schema public revoke all on functions from supabase_admin,public, anon, authenticated;;
alter default privileges in schema public revoke all on sequences from supabase_admin,public, anon, authenticated;;

-- the above will only work for future database objects, to revoke current privileges 
revoke select on all tables in schema public from supabase_admin,public, anon, authenticated;;
revoke usage on all sequences in schema public from supabase_admin,public, anon, authenticated;;
revoke execute on all functions in schema public from supabase_admin,public, anon, authenticated;;
```

(Previously discussed on: https://github.com/supabase/supabase/discussions/1493#discussioncomment-724842)

Once done with the above, you'll need to GRANT specific privileges. Then, column-level security should work:

```sql
GRANT ALL ON todos TO supabase_admin,public, anon, authenticated;

REVOKE UPDATE ("readonly_data") ON todos FROM supabase_admin,public, anon, authenticated;
```

**Reply by mehmetuken** [reply]
```tsql
alter default privileges in schema public revoke all on tables from supabase_admin,public, anon, authenticated;
alter default privileges in schema public revoke all on functions from supabase_admin,public, anon, authenticated;;
alter default privileges in schema public revoke all on sequences from supabase_admin,public, anon, authenticated;;

-- the above will only work for future database objects, to revoke current privileges 
revoke select on all tables in schema public from supabase_admin,public, anon, authenticated;;
revoke usage on all sequences in schema public from supabase_admin,public, anon, authenticated;;
revoke execute on all functions in schema public from supabase_admin,public, anon, authenticated;;

create table todos
(
    id          uuid                     default uuid_generate_v4(),
    user_id     uuid references auth.users                                    not null,
    inserted_at timestamp with time zone DEFAULT timezone('utc'::text, now()) NOT NULL,
    updated_at  timestamp with time zone DEFAULT timezone('utc'::text, now()) NOT NULL,
    name        text,
    readonly_data text,
    primary key (id)
);
alter table todos enable row level security;
create policy "Can only view own todos." on todos for select using (auth.uid() = user_id);
create policy "Can only create own todo." on todos for insert with check(auth.uid() = user_id);
create policy "Can only update own todo." on todos for update using (auth.uid() = user_id);
create policy "Can only delete own todo." on todos for delete using (auth.uid() = user_id);

GRANT ALL ON todos TO supabase_admin,public, anon, authenticated;
REVOKE UPDATE ("readonly_data") ON todos FROM supabase_admin,public, anon, authenticated;
```

my final sql this. it's not working. readonly_data update every time. but i realize i give todos grant update it's 
```tsql
GRANT UPDATE ON todos TO supabase_admin,public, anon, authenticated
```

Api Result
```json
{
    "hint": null,
    "message": "permission denied for table todos",
    "code": "42501",
    "details": null
}
```

but i give only select 
```tsql
GRANT SELECT ON todos TO supabase_admin,public, anon, authenticated
```

it's update work how is it possible?

Api Result
```json
[{
    "id": "6d33e826-53a0-4109-a8ad-3279cd2ad072",
    "user_id": "1570476d-9910-4c9d-8154-74c51116ac9d",
    "inserted_at": "2021-09-12T09:14:16.806234+00:00",
    "updated_at": "2021-09-12T09:14:16.806234+00:00",
    "name": "test 2",
    "readonly_data": "test 2"
}]
```

**Reply by steve-chavez** [reply]
@mehmetuken Right now supabase-js does a select after update. Try using `returning: minimal` to avoid this, check the [docs](https://supabase.io/docs/reference/javascript/update).

```js
await supabase
    .from("todos")
    .update({ readonly_data: "dummy" }, {returning: 'minimal'})
    .eq("id", id)
```

**Reply by mehmetuken** [reply]
@steve-chavez you don't understand me. I do
```tsql
alter default privileges in schema public revoke all on tables from supabase_admin,public, anon, authenticated;
alter default privileges in schema public revoke all on functions from supabase_admin,public, anon, authenticated;;
alter default privileges in schema public revoke all on sequences from supabase_admin,public, anon, authenticated;;

-- the above will only work for future database objects, to revoke current privileges 
revoke select on all tables in schema public from supabase_admin,public, anon, authenticated;;
revoke usage on all sequences in schema public from supabase_admin,public, anon, authenticated;;
revoke execute on all functions in schema public from supabase_admin,public, anon, authenticated;;

GRANT SELECT ON todos TO supabase_admin,public, anon, authenticated;
```

I give just grant select but update it's work. I don't understand this. I'm not giving any update privileges but update method works.

**Reply by steve-chavez** [reply]
Maybe it's because the UPDATE grant hasn't been revoked yet(only SELECT has been revoked). Try adding this line:

```sql
-- revokes SELECT, UPDATE, INSERT, DELETE
revoke all on all tables in schema public from supabase_admin,public, anon, authenticated;;
```

**Reply by mehmetuken** [reply]
```tsql
revoke all on all tables in schema public from supabase_admin,public, anon, authenticated;;
GRANT SELECT ON todos TO supabase_admin,public, anon, authenticated;
```

update working. I think row level security bypass grant or revoke rules.

**Reply by steve-chavez** [reply]
> update working. I think row level security bypass grant or revoke rules.

No, GRANTs are always evaluated first.

Make sure to do the request from the js client, on the SQL Editor the user will indeed bypass GRANTs.

**Reply by mehmetuken** [reply]
Always I'm testing on js client.

**Reply by kiwicopple** [reply]
Hi @mehmetuken - just to confirm: you are using the `anon` key right? Not the `service_role` key?

**Reply by mehmetuken** [reply]
@kiwicopple Yes, i’m using the `anon` key.

**Reply by abhay187** [reply]
We need a solution to this problem. @kiwicopple any update?

**Reply by abdirahmn1** [reply]
Any update on this??,

**Reply by GaryAustin1** [reply]
@abdirahmn1 
You might start a new question this one is very old and confusing all the attempts and what exactly is working/not working.
To use column privileges you first have to take away the table grant for that operation.  Then you have to go back in and add column privileges for all the columns you want to be able to do the operation on.

There is an upcoming UI to help with this: https://supabase.com/blog/supabase-studio-3-0  (see role management)

**Reply by abdirahmn1** [reply]
Thanks @GaryAustin1 for the direction. Your support is appreciated :smile:

**Reply by sanba-anass** [reply]
please how to revert back from this my application stops working

**Reply by GaryAustin1** [reply]
Revert back from what exactly?  What code did you execute?

In general to restore TABLE grants you would run this code:
```
grant usage on schema public to postgres, anon, authenticated, service_role;

grant all privileges on all tables in schema public to postgres, anon, authenticated, service_role;
grant all privileges on all functions in schema public to postgres, anon, authenticated, service_role;
grant all privileges on all sequences in schema public to postgres, anon, authenticated, service_role;

alter default privileges in schema public grant all on tables to postgres, anon, authenticated, service_role;
alter default privileges in schema public grant all on functions to postgres, anon, authenticated, service_role;
alter default privileges in schema public grant all on sequences to postgres, anon, authenticated, service_role;
```
But that assumes you don't have code that removed some table grants that you want to stay that way.  It also likely leaves the column stuff you did (if any) but I THINK they would be ignored.

Unless you know what you are doing messing with table and column grants can get you in trouble.

**Comment by johannessjoberg** [comment]
It seems like `REVOKE UPDATE ("column_x")` has no affect but `REVOKE UPDATE` works as expected. I have no idea why.

I solved this issue by going around it.  First revoke all updates privileges, then grant specific columns.
In your example it would be something like:
```sql
REVOKE UPDATE ON todos FROM supabase_admin,public, anon, authenticated;
GRANT UPDATE("<...all columns but the 'readonly_data'>") ON todos TO supabase_admin,public, anon, authenticated;
```

You can verify the result by checking that the column privilege is not present in `information_schema.column_privileges ` any longer.
```sql
SELECT * FROM information_schema.column_privileges 
WHERE table_name = 'todos' 
AND grantee IN ('supabase_admin', 'public', 'anon', 'authenticated') 
AND privilege_type = 'UPDATE'
AND column_name = 'readonly_data';
```
